ostd/src/sync/spin.rs

```diff
@@ -16,65 +17,87 @@ use crate::{
+
+    fn guard() -> Self::Guard {
+        disable_preempt()
+    }
+}
+
+/// for lock disabling Irq
+pub struct LocalIrqDisabled;
+
+impl Guardian for LocalIrqDisabled {
+    type Guard = DisabledLocalIrqGuard;
+
+    fn guard() -> Self::Guard {
+        disable_local()
+    }
+}
+
+impl<T, G: Guardian> SpinLock<T, G> {
     /// Creates a new spin lock.
     pub const fn new(val: T) -> Self {
-        Self {
+        let lock_inner = SpinLockInner {
             lock: AtomicBool::new(false),
             val: UnsafeCell::new(val),
+        };
+        Self {
+            phantom: PhantomData,
+            inner: lock_inner,
         }
     }
 }
 
-impl<T: ?Sized> SpinLock<T> {
-    /// Acquires the spin lock with disabling the local IRQs. This is the most secure
-    /// locking way.
-    ///
-    /// This method runs in a busy loop until the lock can be acquired.
-    /// After acquiring the spin lock, all interrupts are disabled.
-    pub fn lock_irq_disabled(&self) -> SpinLockGuard<T> {
-        let guard = disable_local();
-        self.acquire_lock();
-        SpinLockGuard_ {
-            lock: self,
-            inner_guard: InnerGuard::IrqGuard(guard),
-        }
-    }
-
-    /// Tries acquiring the spin lock immedidately with disabling the local IRQs.
-    pub fn try_lock_irq_disabled(&self) -> Option<SpinLockGuard<T>> {
-        let irq_guard = disable_local();
-        if self.try_acquire_lock() {
-            let lock_guard = SpinLockGuard_ {
-                lock: self,
-                inner_guard: InnerGuard::IrqGuard(irq_guard),
-            };
-            return Some(lock_guard);
-        }
-        None
+impl<T: ?Sized> SpinLock<T, PreemptDisabled> {
+    /// convert spinlock from PreemptDisabled to LocalIrqDisabled
+    pub fn disable_irq(&self) -> &SpinLock<T, LocalIrqDisabled> {
+        // SAFETY: The memory layout of the source and target types, which only differ in the `phantom` field, are identical because
+        // (1) `PhantomData` is zero-sized, (2) the struct layout follows `repr(C)`.
+        unsafe { core::mem::transmute(self) }
     }
+}
 
+impl<T: ?Sized, G: Guardian> SpinLock<T, G> {
     /// Acquires the spin lock without disabling local IRQs.
     ///
-    /// This method is twice as fast as the [`lock_irq_disabled`] method.
-    /// So prefer using this method over the [`lock_irq_disabled`] method
+    /// This method is twice as fast as the [`disable_irq().lock`] method.
+    /// So prefer using this method over the [`disable_irq().lock`] method
     /// when IRQ handlers are allowed to get executed while
     /// holding this lock. For example, if a lock is never used
     /// in the interrupt context, then it is ok to use this method
     /// in the process context.
     ///
-    /// [`lock_irq_disabled`]: Self::lock_irq_disabled
-    pub fn lock(&self) -> SpinLockGuard<T> {
-        let guard = disable_preempt();
+    /// [`disable_irq().lock`]: Self::disable_irq().lock
```

Comment:
I don't think that link can work. It must point to a method, type, module, etc. But what is `Self::disable_irq().lock`? It should be either `Self::disable_irq` or `Self::lock`.